File: backend/schemas/geo.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional, Dict, Any

class DistrictCrime(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    district: str
    crime_count: int
    latitude: float
    longitude: float

class StationCrime(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    station: str
    crime_count: int
    latitude: float
    longitude: float

class HeatmapPoint(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    latitude: float
    longitude: float
    weight: int

class HotspotCluster(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    cluster_id: int
    crime_count: int
    latitude: float
    longitude: float
    first_incident_date: Optional[str] = None
    last_incident_date: Optional[str] = None
    peak_hour_window: Optional[str] = None
    temporal_category: Optional[str] = "Recurring Historical Cluster"
    hotspot_type: str = "Historical Descriptive Cluster"

class GeoMarkerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    crime_no: str
    crime_type: str
    police_station: str
    district: str
    crime_date: str
    status: str
    latitude: float
    longitude: float

class HourlyDataPoint(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    hour: int
    label: str
    count: int

class TimePeriodDistribution(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    morning: int = 0
    afternoon: int = 0
    evening: int = 0
    night: int = 0

class TimeOfDayResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    hourly: List[HourlyDataPoint]
    periods: TimePeriodDistribution
    category_by_time: List[Dict[str, Any]] = []
    peak_hour: Optional[str] = None
    peak_period: Optional[str] = None
    total_analyzed: int = 0

class GeoIntelligenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    districts: List[DistrictCrime]
    stations: List[StationCrime]
    heatmap: List[HeatmapPoint]
    hotspots: List[HotspotCluster]
    markers: List[GeoMarkerResponse]
    time_of_day: Optional[TimeOfDayResponse] = None

File: frontend/features/geo/types/geo.ts
export interface DistrictCrime {
  district: string;
  crime_count: number;
  latitude: number;
  longitude: number;
}

export interface StationCrime {
  station: string;
  crime_count: number;
  latitude: number;
  longitude: number;
}

export interface HeatmapPoint {
  latitude: number;
  longitude: number;
  weight: number;
}

export interface HotspotCluster {
  cluster_id: number;
  crime_count: number;
  latitude: number;
  longitude: number;
  first_incident_date?: string;
  last_incident_date?: string;
  peak_hour_window?: string;
  temporal_category?: string;
  hotspot_type?: string;
}

export interface GeoMarker {
  id: number;
  crime_no: string;
  crime_type: string;
  police_station: string;
  district: string;
  crime_date: string;
  status: string;
  latitude: number;
  longitude: number;
}

export interface HourlyDataPoint {
  hour: number;
  label: string;
  count: number;
}

export interface TimePeriodDistribution {
  morning: number;
  afternoon: number;
  evening: number;
  night: number;
}

export interface TimeOfDayCategoryBreakdown {
  period: string;
  top_categories: { category: string; count: number }[];
}

export interface TimeOfDayResponse {
  hourly: HourlyDataPoint[];
  periods: TimePeriodDistribution;
  category_by_time: TimeOfDayCategoryBreakdown[];
  peak_hour?: string;
  peak_period?: string;
  total_analyzed: number;
}

export interface GeoFiltersState {
  district?: string;
  police_station?: string;
  crime_type?: string;
  start_date?: string;
  end_date?: string;
<<<<<<< Updated upstream
=======
  time_period?: string;
  /** Minimum crime count per hotspot cluster (controls DBSCAN min_samples sensitivity) */
  min_crime_count?: number;
>>>>>>> Stashed changes
}

export interface GeoLookupOptions {
  districts: string[];
  categories: string[];
  stations: string[];
  stations_by_district: Record<string, string[]>;
}

export interface GeoIntelligenceResponse {
  districts: DistrictCrime[];
  stations: StationCrime[];
  heatmap: HeatmapPoint[];
  hotspots: HotspotCluster[];
  markers: GeoMarker[];
  time_of_day?: TimeOfDayResponse;
}
